tooling/minimize/intrinsics/src/lib.rs

```diff
@@ -29,3 +34,112 @@ pub unsafe fn deallocate(ptr: *mut u8, size: usize, align: usize) {
     let layout = Layout::from_size_align(size, align).unwrap();
     unsafe { System.deallocate(ptr, layout); }
 }
+
+static THREAD_HANDLES: Mutex<Vec<Option<JoinHandle<()>>>> = Mutex::new(Vec::new());
+
+// Spawn returns the index+1 of the thread because we don't have a JoinHandle for thread 0.
+pub fn spawn<F>(f: F) -> usize
+where F: FnOnce() -> () + Send + 'static,
+{
+    let handle = thread::spawn(f);
+    let mut vec = THREAD_HANDLES.lock().unwrap();
+    vec.push(Some(handle));
+    vec.len()
+}
+
+pub fn join(index: usize) {
+    if index == 0 { panic!("Can not join thread 0.") }
+
+    let mut vec = THREAD_HANDLES.lock().unwrap();
+    let join_handle = vec.get_mut(index-1);
+
+    if let Some(join_handle) = join_handle {
+        let join_handle = mem::replace(join_handle, None);
+
+        drop(vec);
+
+        if let Some(join_handle) = join_handle {
+            join_handle.join().unwrap();
+        }
+
+        else {
+            panic!("Joining thread that is joined by other thread.")
+        }
+    }
+    else {
+        panic!("Joining non existent thread.")
+    }
+}
+
+#[derive(Default, PartialEq)]
+enum LockState {
+    Locked,
+    #[default]
+    Unlocked,
+}
+
+static LOCKS: Mutex<Vec<LockState>> = Mutex::new(Vec::new());
+
+pub fn create_lock() -> usize {
+    let mut locks = LOCKS.lock().unwrap();
+    locks.push(LockState::default());
+    locks.len()-1
+}
+
+pub fn acquire(lock_id: usize) {
+    loop {
+        let mut locks = LOCKS.lock().unwrap();
+        if locks.len() <= lock_id {
+            panic!("Acquiring non existent lock.");
+        }
+        let lock = &mut locks[lock_id];
+
+        if *lock == LockState::Unlocked {
+            *lock = LockState::Locked;
+            return;
+        }
+
+        drop(locks);
+
+        thread::park();
+    }
+}
+
+pub fn release(lock_id: usize) {
+    let mut locks = LOCKS.lock().unwrap();
+    if locks.len() <= lock_id {
+        panic!("Releasing non existent lock.");
+    }
+    let lock = &mut locks[lock_id];
+
+    if *lock == LockState::Unlocked {        
+        panic!("Releasing non locked lock.");
+    }
+    *lock = LockState::Unlocked;
+    drop(locks);
+
+    let thread_handles = THREAD_HANDLES.lock().unwrap();
+    for handle in thread_handles.as_slice() {
+        if let Some(handle) = handle {
+            handle.thread().unpark();
```

Comment:
Uh what is going on here?^^

I suggest you reuse the locks from the standard library instead of implementing your own. (Also, unparking *all threads* is a bit too much.^^ A custom implementation should store a queue of waiting threads in the `Locked` state.)

Reply:
Will add a comment to the code explaining why there is my own lock implementation.